Add Nesterov Momentum

ajbrock opened the pull request:
This pull request adds the Nesterov Momentum option to SGD, using the parameterization described [here](https://github.com/lisa-lab/pylearn2/pull/136#issuecomment-10381617). I'm not sure if any of the tests need to be modified to include this, but the code here should be correct (tested and working on my CIFAR100 testbed).

ajbrock committed:
Add Nesterov Momentum

Adds the Nesterov variant of momentum as an option for SGD.

ajbrock committed:
Fix Mistake in missing "defaults" dict arg

derp

ajbrock committed:
Add appropriate whitespace

ajbrock committed:
Multiply by momentum before adding to d_p

apaszke commented on torch/optim/sgd.py:
```diff
@@ -51,7 +51,10 @@ def step(self, closure=None):
                     else:
                         buf = param_state['momentum_buffer']
-                        d_p = buf.mul_(momentum).add_(1 - dampening, d_p)
+                        if nesterov !=0:
```
How about this:
```python
buf.mul_(momentum).add_(1 - dampening, d_p)
if nesterov:
    d_p.add_(momentum, buf)
else:
    d_p = buf
```

ajbrock replied:
Looks good, I'll change it to this.

apaszke commented on torch/optim/sgd.py:
```diff
@@ -3,31 +3,30 @@
 
     def __init__(self, params, lr=required, momentum=0, dampening=0,
-                 weight_decay=0):
+                 weight_decay=0, nesterov=False):
```
Can you please add a check that if `nesterov` is True, then `momentum` has be be greater than 0?

ajbrock replied:
Should this check also include a condition that dampening be zero, as in the [legacy](https://github.com/ajbrock/pytorch/blob/master/torch/legacy/optim/sgd.py#L44) version?

apaszke reviewed (commented):
There's one more problem - serialized state dicts or optimizers won't have the `nesterov` attribute in any parameter group. I'll figure out how to solve this tomorrow, and I'll get back to you then. Thanks for the PR!

ajbrock committed:
Update sgd.py

ajbrock committed:
Update sgd.py

ajbrock committed:
Replace F.Linear calls with non-stateless methods

This replaces the calls to mm and mv in F.Linear with stateless methods, so that training with HalfTensors with a linear output works.

soumith committed:
Update sgd.py

soumith commented:
@pytorchbot test this

ajbrock committed:
undo accidental commit to wrong branch

Accidentally committed to the Nesterov Momentum PR

apaszke commented:
Pasted from #810:

Ok, so I thought this through and I think the best way would be to implement __setstate__ such that it calls to super and then patches the param groups like that:

```python
def __setstate__(self, state):
    super(RMSprop, self).__setstate__(state)
    for group in self.parameter_groups:
        group.set_default('momentum', 0)
        group.set_default('centered', False)
```

Sorry for waiting!

apaszke committed:
Add __setstate__ to SGD

apaszke commented:
I added the `__setstate__` and opened a new PR: #887. Thanks!
